home/tests/test_api.py

```diff
@@ -244,118 +254,142 @@ def test_pagination(self):
+    def test_detail_view_with_variations(self, uclient):
+        """
+        FIXME:
+         * It's unclear what this is actually testing.
+         * This should probably be in a class without the content setup.
+        """
         ContentPage.objects.all().delete()
-        self.assertEqual(PageView.objects.count(), 0)
+        assert PageView.objects.count() == 0
 
         # variations should be in the whatsapp content
         page = create_page(tags=["tag1", "tag2"], add_variation=True)
 
-        response = self.client.get(f"/api/v2/pages/{page.id}/?whatsapp=true&message=1")
+        response = uclient.get(f"/api/v2/pages/{page.id}/?whatsapp=true&message=1")
         content = response.json()
 
         var_content = content["body"]["text"]["value"]["variation_messages"]
-        self.assertEqual(1, len(var_content))
-        self.assertEqual(var_content[0]["profile_field"], "gender")
-        self.assertEqual(var_content[0]["value"], "female")
-        self.assertEqual(var_content[0]["message"], "Test Title - female variation")
+        assert len(var_content) == 1
+        assert var_content[0]["profile_field"] == "gender"
+        assert var_content[0]["value"] == "female"
+        assert var_content[0]["message"] == "Test Title - female variation"
 
-        self.assertEqual(PageView.objects.count(), 1)
+        assert PageView.objects.count() == 1
         view = PageView.objects.last()
-        self.assertEqual(view.message, 1)
+        assert view.message == 1
 
-    def test_whatsapp_body(self):
+    def test_whatsapp_body(self, uclient):
         """
         Should have the WhatsApp specific fields included in the body; if it's a
         template, what's the template name, the text body of the message.
+
+        FIXME:
+         * This should probably be in a class without the content setup.
         """
         ContentPage.objects.all().delete()
         page = create_page(
             is_whatsapp_template=True, whatsapp_template_name="test_template"
         )
 
         # it should return the correct details
-        response = self.client.get(f"/api/v2/pages/{page.id}/?whatsapp")
+        response = uclient.get(f"/api/v2/pages/{page.id}/?whatsapp")
         content = response.json()
-        self.assertTrue(content["body"]["is_whatsapp_template"])
-        self.assertEqual(content["body"]["whatsapp_template_name"], "test_template")
-        self.assertEqual(
-            content["body"]["text"]["value"]["message"], "Test WhatsApp Message 1"
-        )
+        assert content["body"]["is_whatsapp_template"]
+        assert content["body"]["whatsapp_template_name"] == "test_template"
+        assert content["body"]["text"]["value"]["message"] == "Test WhatsApp Message 1"
 
-    def test_detail_view_no_content_page(self):
+    def test_detail_view_no_content_page(self, uclient):
+        """
+        FIXME:
+         * It's unclear what the intended result is.
+         * Should the response code be a success?
```

Comment:
I'm unsure whether it makes sense to be a success or not. I've seen some good arguments not to return a 404 in REST APIs if an object with that key doesn't exist, because that doesn't tell you if you've got the URL path wrong, or if it's an object that doesn't exist.

Either way, we should probably assert on the response code to make it clear what we expect.

Reply:
Proper REST APIs should be giving you resource URLs instead of making you construct them yourself, so getting the structure wrong shouldn't be a concern there. 🤷 I think it's safe to assume that the client is either going to be requesting correct URLs (that may point to things that don't exist) in which case 404 for "the thing you asked for isn't here" makes sense, or it's so broken that differentiating between "not found" and "nonsensical request" won't actually help at all.

I'll add an assert on the response code and maybe a comment if it's not 404.

Reply:
Looks like we explicitly catch `ContentPage.DoesNotExist` (presumably 404) and raise `ValidationError` (400) instead. This definitely needs a rethink. 😕

Reply:
Hmm, looks to be that we don't want a 500 error on missing IDs, which is correct, and would happen if we try to save a page view for a ContentPage that doesn't exist. But then we avoid that by raising a `ValidationError`, which is a 400 error, which doesn't seem correct to me..

Reply:
We can raise `rest_framework.exceptions.NotFound` instead of `rest_framework.exceptions.ValidationError`, but that's an API change. Something for the design doc or a separate issue, I guess. 🤷